README.md
# Rainbow

[![PkgGoDev](https://pkg.go.dev/badge/github.com/arsham/rainbow)](https://pkg.go.dev/github.com/arsham/rainbow)
![GitHub go.mod Go version](https://img.shields.io/github/go-mod/go-version/arsham/rainbow)
[![Build Status](https://travis-ci.org/arsham/rainbow.svg?branch=master)](https://travis-ci.org/arsham/rainbow)
[![Coverage Status](https://codecov.io/gh/arsham/rainbow/branch/master/graph/badge.svg)](https://codecov.io/gh/arsham/rainbow)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/9d181f84a8ab4ab3a07201e37d9a218b)](https://www.codacy.com/app/arsham/rainbow?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=arsham/rainbow&amp;utm_campaign=Badge_Grade)
[![Go Report Card](https://goreportcard.com/badge/github.com/arsham/rainbow)](https://goreportcard.com/report/github.com/arsham/rainbow)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)


Tasty rainbows for your terminal like these:

![Screenshot](/docs/rainbow.png?raw=true "Rainbow")

This app was inspired by lolcats, but written in Go.

### Table of Contents

1. [Installation](#installation)
2. [Usage](#usage)
3. [As library](#as-library)
4. [See Also](#see-also)
5. [License](#license)

## Installation

Get the library:
```bash
go get github.com/arsham/rainbow
```

## Usage

You can pipe the text into the app in many ways. Choose one that is suitable for
you:

```bash
# File contents:
rainbow < filename.txt

# Echo a string:
echo "Any quotes" | rainbow

# Here string:
rainbow <<END
Consectetur aliqua do quis sed
proident enim fugiat occaecat nisi
in deserunt culpa aliquip do excepteur.
END

# Output of a program:
ls -l | rainbow
```

## As library

`Light` struct implements io.Reader and io.Writer:

```go
import "github.com/arsham/rainbow/rainbow"
// ...
l := rainbow.Light{
    Reader: someReader, // to read from
    Writer: os.Stdout, // to write to
}
l.Paint() // will rainbow everything it reads from reader to writer.
```

If you want the rainbow to be random, you can seed it this way:

```go
l := rainbow.Light{
    Reader: buf,
    Writer: os.Stdout,
    Seed:   rand.Int63n(256),
}
```

Or if you prefer io.Copy:

```go
l := rainbow.Light{
    Writer: os.Stdout,
    Seed:   rand.Int63n(256),
}
io.Copy(l, someReader)
```

## See Also

See also [Figurine][figurine]. It prints beautiful ASCII arts in FIGlet.

## License
Use of this source code is governed by the Apache 2.0 license. License that can
be found in the [LICENSE](./LICENSE) file.

Please note that this was initially forked from [glolcat][glolcat], but I
decided to rewrite it as the library is not maintained anymore.

[figurine]: https://github.com/arsham/figurine
[glolcat]: https://github.com/cezarsa/glolcat


CHANGELOG.md
# CHANGELOG

<a name="unreleased"></a>
## [Unreleased]



<a name="v1.1.1"></a>
## [v1.1.1] - 2020-08-18

### Performance Improvements
- Performance improvements over cpu and memory


<a name="v1.1.0"></a>
## [v1.1.0] - 2020-08-16

### Code Refactoring
- Refactor code base and performance improvements


<a name="v1.0.1"></a>
## [v1.0.1] - 2019-12-01

### Added
- Add error checks in tests

### Fixed
- Fix release names


<a name="v1.0.0"></a>
## [v1.0.0] - 2019-06-17

### Replace
- Replace glide with go modules


<a name="v0.1"></a>
## [v0.1] - 2018-10-16

### Added
- add Makefile update README.md file
- add utf8 support (closes [#2](https://github.com/arsham/rainbow/issues/2))

### Use
- use github.com/arsham/strings for performance improvements


<a name="Rainbow"></a>
## Rainbow - 2018-04-20

### Added
- add coverage script update travis script
- add README.md and LICENSE files receive input from arguments and shell pipes add tests and documentations

### Initial
- Initial commit



[Unreleased]: https://github.com/arsham/rainbow/compare/v1.1.1...HEAD
[v1.1.1]: https://github.com/arsham/rainbow/compare/v1.1.0...v1.1.1
[v1.1.0]: https://github.com/arsham/rainbow/compare/v1.0.1...v1.1.0
[v1.0.1]: https://github.com/arsham/rainbow/compare/v1.0.0...v1.0.1
[v1.0.0]: https://github.com/arsham/rainbow/compare/v0.1...v1.0.0
[v0.1]: https://github.com/arsham/rainbow/compare/Rainbow...v0.1


main_test.go
// Copyright 2016 Arsham Shirvani <arshamshirvani@gmail.com>. All rights reserved.
// Use of this source code is governed by the Apache 2.0 license
// License that can be found in the LICENSE file.

package main

import (
	"bytes"
	"io/ioutil"
	"os"
	"regexp"
	"testing"

	"github.com/brianvoe/gofakeit/v5"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

var re = regexp.MustCompile(`\x1B\[[0-9;]*[JKmsu]`)

func setup(t *testing.T) func() {
	t.Helper()
	oldStdin := os.Stdin
	oldStdout := os.Stdout
	oldArgs := os.Args

	fin, err := ioutil.TempFile("", "testMain")
	require.NoError(t, err)
	fout, err := ioutil.TempFile("", "testMain")
	if err != nil {
		assert.NoError(t, fin.Close())
		t.Fatal(err)
	}
	os.Stdin = fin
	os.Stdout = fout
	return func() {
		os.Stdin = oldStdin
		os.Stdout = oldStdout
		os.Args = oldArgs
		assert.NoError(t, fin.Close())
		assert.NoError(t, fout.Close())
		assert.NoError(t, os.Remove(fin.Name()))
		assert.NoError(t, os.Remove(fout.Name()))
	}
}

func TestMain(t *testing.T) {
	t.Run("WithArgs", testMainWithArgs)
	t.Run("WithPipe", testMainWithPipe)
	t.Run("CopyError", testMainCopyError)
}

func testMainWithArgs(t *testing.T) {
	cleanup := setup(t)
	defer cleanup()
	input := gofakeit.Sentence(20)
	os.Args = []string{"rainbow", input}
	main()
	os.Stdout.Seek(0, 0)
	buf := &bytes.Buffer{}
	buf.ReadFrom(os.Stdout)

	out := buf.Bytes()
	got := re.ReplaceAll(out, []byte(""))
	assert.Equal(t, []byte(input+"\n"), got)
}

func testMainWithPipe(t *testing.T) {
	cleanup := setup(t)
	defer cleanup()
	input := gofakeit.Sentence(20)
	os.Args = []string{"rainbow"}
	os.Stdin.WriteString(input)
	os.Stdin.Seek(0, 0)
	main()
	os.Stdout.Seek(0, 0)
	buf := &bytes.Buffer{}
	buf.ReadFrom(os.Stdout)

	out := buf.Bytes()
	got := re.ReplaceAll(out, []byte(""))
	assert.Equal(t, []byte(input), got)
}

func testMainCopyError(t *testing.T) {
	cleanup := setup(t)
	defer cleanup()
	fin, err := ioutil.TempFile("", "testMain")
	require.NoError(t, err)
	require.NoError(t, fin.Close())
	os.Stdin = fin

	os.Args = []string{"rainbow"}
	main()
	os.Stdout.Seek(0, 0)
	buf := &bytes.Buffer{}
	buf.ReadFrom(os.Stdout)
	got := buf.String()
	assert.Contains(t, got, "already closed")
}


app/app_test.go
// Copyright 2016 Arsham Shirvani <arshamshirvani@gmail.com>. All rights reserved.
// Use of this source code is governed by the Apache 2.0 license
// License that can be found in the LICENSE file.

package app_test

import (
	"bytes"
	"io/ioutil"
	"os"
	"regexp"
	"testing"

	"github.com/arsham/rainbow/app"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

var re = regexp.MustCompile(`\x1B\[[0-9;]*[JKmsu]`)

func setup(t *testing.T) func() {
	t.Helper()
	oldStdin := os.Stdin
	oldStdout := os.Stdout
	oldArgs := os.Args

	fin, err := ioutil.TempFile("", "testMain")
	require.NoError(t, err)
	fout, err := ioutil.TempFile("", "testMain")
	if err != nil {
		assert.NoError(t, fin.Close())
		t.Fatal(err)
	}
	os.Stdin = fin
	os.Stdout = fout
	return func() {
		os.Stdin = oldStdin
		os.Stdout = oldStdout
		os.Args = oldArgs
		assert.NoError(t, fin.Close())
		assert.NoError(t, fout.Close())
		assert.NoError(t, os.Remove(fin.Name()))
		assert.NoError(t, os.Remove(fout.Name()))
	}
}

func TestMainArg(t *testing.T) {
	cleanup := setup(t)
	defer cleanup()
	input := "RjAuFZXfyIkaE5Ox LitwdlBD6E0GL4Y6p"
	os.Args = []string{"rainbow", input}
	app.Main()
	os.Stdout.Seek(0, 0)
	buf := new(bytes.Buffer)
	buf.ReadFrom(os.Stdout)

	out := buf.Bytes()
	got := re.ReplaceAll(out, []byte(""))
	assert.Equal(t, []byte(input+"\n"), got)
}

func TestMainPipe(t *testing.T) {
	cleanup := setup(t)
	defer cleanup()
	input := "RjAuFZXfyIkaE5Ox LitwdlBD6E0GL4Y6p"
	os.Args = []string{"rainbow"}
	os.Stdin.WriteString(input)
	os.Stdin.Seek(0, 0)
	app.Main()
	os.Stdout.Seek(0, 0)
	buf := new(bytes.Buffer)
	buf.ReadFrom(os.Stdout)

	out := buf.Bytes()
	got := re.ReplaceAll(out, []byte(""))
	assert.Equal(t, []byte(input), got)
}


rainbow/rainbow_test.go
// Copyright 2016 Arsham Shirvani <arshamshirvani@gmail.com>. All rights reserved.
// Use of this source code is governed by the Apache 2.0 license
// License that can be found in the LICENSE file.

package rainbow

import (
	"bytes"
	"errors"
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"regexp"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func init() {
	rand.Seed(time.Now().UTC().UnixNano())
}

func readFile(t *testing.T, name string) []byte {
	f, err := os.Open("testdata/" + name)
	require.NoError(t, err)
	b, err := ioutil.ReadAll(f)
	require.NoError(t, err)
	return b
}

func TestLightPaint(t *testing.T) {
	t.Parallel()
	plain := string(readFile(t, "plain.txt"))
	painted := string(readFile(t, "painted.txt"))
	//nolint:stylecheck // this is on purpose.
	tcs := []struct {
		sample  string
		painted string
	}{
		{
			"2d7gMRSgGLj9F0c tPjSmsdRsTej4x7BJiOp R9HUHEiyH0G1Ld XeL5fjQ1KkxI3",
			`[38;5;154m2[0m[38;5;154md[0m[38;5;154m7[0m[38;5;154mg[0m[38;5;154mM[0m[38;5;154mR[0m[38;5;154mS[0m[38;5;148mg[0m[38;5;184mG[0m[38;5;184mL[0m[38;5;184mj[0m[38;5;184m9[0m[38;5;184mF[0m[38;5;184m0[0m[38;5;184mc[0m[38;5;184m [0m[38;5;184mt[0m[38;5;184mP[0m[38;5;184mj[0m[38;5;178mS[0m[38;5;214mm[0m[38;5;214ms[0m[38;5;214md[0m[38;5;214mR[0m[38;5;214ms[0m[38;5;214mT[0m[38;5;214me[0m[38;5;214mj[0m[38;5;214m4[0m[38;5;208mx[0m[38;5;208m7[0m[38;5;208mB[0m[38;5;208mJ[0m[38;5;208mi[0m[38;5;208mO[0m[38;5;208mp[0m[38;5;208m [0m[38;5;208mR[0m[38;5;209m9[0m[38;5;203mH[0m[38;5;203mU[0m[38;5;203mH[0m[38;5;203mE[0m[38;5;203mi[0m[38;5;203my[0m[38;5;203mH[0m[38;5;203m0[0m[38;5;203mG[0m[38;5;203m1[0m[38;5;203mL[0m[38;5;204md[0m[38;5;198m [0m[38;5;198mX[0m[38;5;198me[0m[38;5;198mL[0m[38;5;198m5[0m[38;5;198mf[0m[38;5;198mj[0m[38;5;198mQ[0m[38;5;198m1[0m[38;5;199mK[0m[38;5;199mk[0m[38;5;199mx[0m[38;5;199mI[0m[38;5;199m3[0m`,
		},
		{
			"11✂1",
			`[38;5;154m1[0m[38;5;154m1[0m[38;5;154m✂[0m[38;5;154m1[0m`,
		},
		{
			"🏧-✂1",
			`[38;5;154m🏧[0m[38;5;154m-[0m[38;5;154m✂[0m[38;5;154m1[0m`,
		},
		{
			plain,
			painted,
		},
	}
	for _, tc := range tcs {
		r := strings.NewReader(tc.sample)
		w := &bytes.Buffer{}
		l := &Light{
			Reader: r,
			Writer: w,
			Seed:   1,
		}
		err := l.Paint()
		assert.NoError(t, err)
		assert.EqualValues(t, tc.painted, w.String())
	}
}

func BenchmarkLightPaint(b *testing.B) {
	bcs := []struct {
		lines   int
		letters int
	}{
		{1, 10},
		{1, 240},
		{10, 10},
		{10, 240},
		{100, 10},
		{100, 240},
		{500, 500},
	}
	for _, bc := range bcs {
		var (
			totalLen int
			name     = fmt.Sprintf("lines%d_let%d", bc.lines, bc.letters)
			line     = make([]byte, bc.letters)
			r        = &bytes.Buffer{}
			w        = &bytes.Buffer{}
		)
		rand.Read(line)
		for i := 0; i < bc.lines; i++ {
			r.Write(line)
			r.Write([]byte("\n"))
			totalLen += len(line) + 1
		}
		b.ResetTimer()
		b.Run(name, func(b *testing.B) {
			b.Run("Serial", func(b *testing.B) {
				b.ResetTimer()
				l := &Light{
					Writer: w,
					Reader: r,
					Seed:   1,
				}
				for i := 0; i < b.N; i++ {
					l.Paint()
				}
			})
			b.Run("Parallel", func(b *testing.B) {
				b.ResetTimer()
				b.RunParallel(func(bp *testing.PB) {
					l := &Light{
						Writer: w,
						Reader: r,
						Seed:   1,
					}
					for bp.Next() {
						l.Paint()
					}
				})
			})
		})
	}
}

func TestPlotPos(t *testing.T) {
	t.Parallel()
	tcs := []struct {
		name  string
		x     float64
		red   int
		green int
		blue  int
	}{
		{"0", 0, 128, 237, 18},
		{"1", 1, 140, 231, 12},
		{"5", 5, 188, 194, 1},
		{"10", 10, 234, 133, 15},
		{"50", 50, 6, 220, 157},
		{"100", 100, 58, 70, 254},
		{"360", 360, 2, 176, 205},
	}
	for _, tc := range tcs {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			got, got1, got2 := plotPos(tc.x)
			assert.Equal(t, tc.red, int(got), "red value")
			assert.Equal(t, tc.green, int(got1), "green value")
			assert.Equal(t, tc.blue, int(got2), " blue value")
		})
	}
}

var got, got1, got2 float64

func BenchmarkPlotPos(b *testing.B) {
	b.Run("Serial", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			got, got1, got2 = plotPos(100)
		}
	})
	b.Run("Parallel", func(b *testing.B) {
		b.RunParallel(func(b *testing.PB) {
			for b.Next() {
				got, got1, got2 = plotPos(100)
			}
		})
	})
}

// this test is here for keeping the logic in sync when we refactor the codes.
func TestColour(t *testing.T) {
	t.Parallel()
	randColour := func() int32 {
		return rand.Int31n(256)
	}
	bc := func(value float64, factor int) int {
		return int(6*value/256) * factor
	}
	check := func(red, green, blue float64) int {
		return 16 + bc(red, 36) + bc(green, 6) + bc(blue, 1)
	}
	for i := 0; i < 1000; i++ {
		red, green, blue := float64(randColour()), float64(randColour()), float64(randColour())
		got := colour(red, green, blue)
		want := check(red, green, blue)
		assert.Equalf(t, want, got, "colour(%f, %f, %f)", red, green, blue)
	}
}

var intval int

func BenchmarkColour(b *testing.B) {
	b.Run("Serial", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			intval = colour(6, 100, 1000)
		}
	})
	b.Run("Parallel", func(b *testing.B) {
		b.RunParallel(func(b *testing.PB) {
			for b.Next() {
				intval = colour(6, 100, 1000)
			}
		})
	})
}

type writeError func([]byte) (int, error)

func (w *writeError) Write(p []byte) (int, error) { return (*w)(p) }

func TestLightWrite(t *testing.T) {
	errExam := errors.New("this error")
	wrErr := writeError(func([]byte) (int, error) { return 0, errExam })
	//nolint:stylecheck // this is on purpose.
	tcs := map[string]struct {
		data     []byte
		want     []byte
		checkErr bool
	}{
		"new line":           {[]byte("\n"), []byte("\n"), true},
		"tab":                {[]byte("\t"), []byte("\t"), true},
		"NL tab":             {[]byte("\n\t"), []byte("\n\t"), true},
		"tab NL":             {[]byte("\t\n"), []byte("\t\n"), true},
		`033[38;5;2m`:        {[]byte("\033[38;5;2m"), []byte(""), false},
		`033[38;5;2K`:        {[]byte("\033[38;5;2K"), []byte(""), false},
		`033[32K`:            {[]byte("\033[32K"), []byte(""), false},
		`033[3K`:             {[]byte("\033[3K"), []byte(""), false},
		`033[3KARSHAM bytes`: {[]byte("\033[3KARSHAM"), []byte{27, 91, 51, 56, 59, 53, 59, 49, 53, 52, 109, 65, 27, 91, 48, 109, 27, 91, 51, 56, 59, 53, 59, 49, 53, 52, 109, 82, 27, 91, 48, 109, 27, 91, 51, 56, 59, 53, 59, 49, 53, 52, 109, 83, 27, 91, 48, 109, 27, 91, 51, 56, 59, 53, 59, 49, 53, 52, 109, 72, 27, 91, 48, 109, 27, 91, 51, 56, 59, 53, 59, 49, 53, 52, 109, 65, 27, 91, 48, 109, 27, 91, 51, 56, 59, 53, 59, 49, 53, 52, 109, 77, 27, 91, 48, 109}, true},
		`033[3KARSHAM string`: {[]byte("\033[3KARSHAM"), []byte("[38;5;154mA[0m[38;5;154mR[0m[38;5;154mS[0m[38;5;154mH[0m[38;5;154mA[0m[38;5;154mM[0m"), true},
	}
	for name, tc := range tcs {
		tc := tc
		t.Run(name, func(t *testing.T) {
			buf := &bytes.Buffer{}
			l := &Light{
				Writer: buf,
				Seed:   1,
			}
			_, err := l.Write(tc.data)
			assert.NoError(t, err)
			got := buf.Bytes()
			if !bytes.Equal(got, tc.want) {
				t.Errorf("got (%v), want (%v)", got, tc.want)
			}
			if !tc.checkErr {
				return
			}
			l.Writer = &wrErr
			_, err = l.Write(tc.data)
			assert.Error(t, err)
		})
	}
	r := &bytes.Buffer{}
	l := &Light{
		Reader: r,
	}
	n, err := l.Write([]byte("blah"))
	assert.Error(t, err)
	assert.Zero(t, n)
}

func TestLightWriteRace(t *testing.T) {
	var (
		wg    sync.WaitGroup
		count = 1000
		data  = bytes.Repeat([]byte("abc def\n"), 10)
		l     = &Light{
			Writer: ioutil.Discard,
			Seed:   1,
		}
	)
	wg.Add(3)
	go func() {
		for i := 0; i < count; i++ {
			l.Write(data)
		}
		wg.Done()
	}()
	go func() {
		for i := 0; i < count; i++ {
			l.Write(data)
		}
		wg.Done()
	}()
	go func() {
		for i := 0; i < count; i++ {
			l.Write(data)
		}
		wg.Done()
	}()
	wg.Wait()
}

func BenchmarkLightWrite(b *testing.B) {
	bcs := []struct {
		line   int
		length int
		data   string
	}{
		{1, 1, "\n"},
		{5, 10, strings.Repeat("aaaaa\n", 10)},
		{15, 50, strings.Repeat("aaaaabbbbbccccc\n", 15)},
		{15, 100, strings.Repeat(strings.Repeat("abcde", 20)+"\n", 15)},
		{50, 50, strings.Repeat(strings.Repeat("abcde", 10)+"\n", 50)},
		{100, 120, strings.Repeat(strings.Repeat("a", 120)+"\n", 100)},
	}
	b.ResetTimer()
	b.Run("Serial", func(b *testing.B) {
		for _, bc := range bcs {
			bc := bc
			l := &Light{
				Writer: ioutil.Discard,
				Seed:   1,
			}
			name := fmt.Sprintf("line%d_len%d", bc.line, bc.length)
			b.Run(name, func(b *testing.B) {
				b.ResetTimer()
				for i := 0; i < b.N; i++ {
					l.Write([]byte(bc.data))
				}
			})
		}
	})
	b.Run("Parallel", func(b *testing.B) {
		for _, bc := range bcs {
			bc := bc
			l := &Light{
				Writer: ioutil.Discard,
				Seed:   1,
			}
			name := fmt.Sprintf("line%d_len%d", bc.line, bc.length)
			b.Run(name, func(b *testing.B) {
				b.ResetTimer()
				b.RunParallel(func(b *testing.PB) {
					for b.Next() {
						l.Write([]byte(bc.data))
					}
				})
			})
		}
	})
}

// making sure we are not altering any texts.
func TestLightWriteRevert(t *testing.T) {
	re := regexp.MustCompile(`\x1B\[[0-9;]*[JKmsu]`)
	tcs := []string{
		"u9VGCQ1E4KCr8bO8 3ULdtlHL3WsjulJU kqUneSFT6 tvyAfih1Qew 5wBKffL4Yc",
		"Y5LFQNulLC0GTKB W4buVQmQTMu6C7aFs uGL6 x2OgVRlUZHCq46kgk sjr 4HKIb",
		"MamsTagRix6bEYwBGR b9FK7b1L 5x1YtTo8nFLrz0dZ rIdZdY0 b0bC05T42bHfV",
		"Ubmba7pnOCoUw9 xGgDjSIPU 7vOwUoiPHeCoxT XtywrjciBYZR cBQySayKWb rx",
		"EPz0kb3pQUuVv LL0t4t8mNaRklyuZTPi wHI2H35IReZbbdb9akXw gMCuJ PRrVK",
		"9qR4HkZ86 enPzoAyIWfz3bFg 8LSokcvV47J0 XfHM ngASr MfkM53zAkwLZmY  ",
		"G7R9mjoYz9FU306 ATfgJ2C6AokIgtf BmL5uLWSJFTDs VO85P JXUjV 5n4OEuvl",
		"32Tj lwI9YaHDjsbqZUBIi 5XiJ7tOh 4eCqaaHT1i WpsJh3JA7s HKHk5R9yPKjL",
		"188qrmdf0GLU9 E2g vp5iX4g2CJ ueKSPvwY369daXZU 0bhg7IGhzeegWeGk5Fj ",
		"YXXUN751 qkTsoR95Udu EtFAN NgBAEe97uzp wpu2VKcX0W20P084d V1LLH28Rk",
		"nSZ6 Qww4GOxAsduJkEPIuNln GGbS rNq YEfhc4jaTAzoHC eT1ugr0mMYANX5JX",
		"wFZkZVWqV3ag7pYtlyN mKtPoVvvZMgU3p6E1 u6zSuUFHuk8nZsvQQ5 qeq 14YPo",
		"uXDLM l2J sznWecxn Ayjv9Ii3akvRD ArTwAkrA THyyrqT6LSAGnfJSSMx4Mes ",
		"m3Lo ufeV1XmoFxqs 4LE21GCiI5UT51 pj9YV B2a43pJxjCg b5CqaspplX5N2dq",
		"9BwGLsCTNGXDfLXNl ZcPNImZhzmDp8S 3ZG177TSj tjOSRMxBZ rhjP0zwJU1K o",
		"4MySJAF rybjdvOUuZhf VqPKvVuw5jaNsldI 8oYI8ZTL2s mZ7X 4awf4PPeLIHh",
		"LTLj3ayz83gM N5So9T32GVipPB2B Ccy 8UutGx N7u6DeZ bUN1hsBDoSC4Z0 sd",
		"gybKEl1 73H5qkvjR UbTUZl jAeMVFeDCtAGVbGhCq Fi0ZQctnmhtk0edD5 gkU3",
		"YX4dsZWyblp6PJIqTR vTgLmrYZd3 MlnwDVOtS1wKZDpoqlxY vu0ivZaYrqWhzAV",
		"Vb1Nks0QpBl1CTkJF QiO7nMLehiI0u S2XwbQXs9Znz Mbe0BQ13JTAOmSjmh WFH",
		"GNIzszrfT5 8WBHpE00a5j7Srfnx e8Qrrhomy8tw7XIa kQFG7ZazYio x5z PZIQ",
	}
	for _, tc := range tcs {
		w := &bytes.Buffer{}
		r := bytes.NewReader([]byte(tc))
		l := Light{
			Reader: r,
			Writer: w,
		}
		err := l.Paint()
		require.NoError(t, err)

		got := re.ReplaceAll(w.Bytes(), []byte(""))
		assert.EqualValues(t, tc, got)
	}
}


rainbow/rainbow.go
// Copyright 2020 Arsham Shirvani <arshamshirvani@gmail.com>. All rights reserved.
// Use of this source code is governed by the Apache 2.0 license
// License that can be found in the LICENSE file.

// Package rainbow prints texts in beautiful rainbows in terminal. Usage is very
// simple:
//
//   import "github.com/arsham/rainbow/rainbow"
//   // ...
//   l := rainbow.Light{
//       Reader: someReader, // to read from
//       Writer: os.Stdout, // to write to
//   }
//   l.Paint() // will rainbow everything it reads from reader to writer.
//
// If you want the rainbow to be random, you can seed it this way:
//   l := rainbow.Light{
//       Reader: buf,
//       Writer: os.Stdout,
//       Seed:   rand.Int31n(256),
//   }
//
// You can also use the Light as a Writer:
//   l := rainbow.Light{
//       Writer: os.Stdout, // to write to
//       Seed:   rand.Int31n(256),
//   }
//   io.Copy(l, someReader)
package rainbow

import (
	"bytes"
	"errors"
	"io"
	"math"
	"math/rand"
	"regexp"
	"strconv"
)

var (
	// we remove all previous paintings to create a new rainbow.
	colorMatch = regexp.MustCompile("^\033" + `\[\d+(;\d+)?(;\d+)?[mK]`)

	// ErrNilWriter is returned when Light.Writer is nil.
	ErrNilWriter = errors.New("nil writer")
)

const (
	freq   = 0.1
	spread = 3
)

// Light reads data from the Writer and pains the contents to the Reader. You
// should seed it everytime otherwise you get the same results.
type Light struct {
	Reader io.Reader
	Writer io.Writer
	Seed   int64
}

// Paint returns an error if it could not copy the data.
func (l *Light) Paint() error {
	if l.Seed == 0 {
		l.Seed = rand.Int63n(256)
	}
	_, err := io.Copy(l, l.Reader)
	return err
}

// Write paints the data and writes it into l.Writer.
func (l *Light) Write(data []byte) (int, error) {
	if l.Writer == nil {
		return 0, ErrNilWriter
	}
	var (
		offset  float64
		dataLen = len(data)
		// 16 times seems to be the sweet spot.
		buf  = bytes.NewBuffer(make([]byte, 0, dataLen*16))
		seed = l.Seed
	)

	data = colorMatch.ReplaceAll(data, []byte(""))
	for _, c := range string(data) {
		switch c {
		case '\n':
			offset = 0
			seed++
			buf.WriteByte('\n')
		case '\t':
			offset++
			buf.WriteByte('\t')
		default:
			r, g, b := plotPos(float64(seed) + (offset / spread))
			colouriseWriter(buf, c, r, g, b)
			offset++
		}
	}
	_, err := l.Writer.Write(buf.Bytes())
	return dataLen, err
}

func plotPos(x float64) (red, green, blue float64) {
	red = math.Sin(freq*x)*127 + 128
	green = math.Sin(freq*x+2*math.Pi/3)*127 + 128
	blue = math.Sin(freq*x+4*math.Pi/3)*127 + 128
	return red, green, blue
}

const max = 16 + (6 * (127 + 128) / 256 * 36) + (6 * (127 + 128) / 256 * 6) + (6 * (127 + 128) / 256)

// nums is used to cache the values of strconv.Itoa(n) for better performance
// gains.
var nums = make([]string, 0, max)

func init() {
	for i := 0; i < max; i++ {
		nums = append(nums, strconv.Itoa(i))
	}
}

func colouriseWriter(s *bytes.Buffer, c rune, r, g, b float64) {
	s.WriteString("\033[38;5;")
	s.WriteString(nums[colour(r, g, b)])
	s.WriteByte('m')
	s.WriteRune(c)
	s.WriteString("\033[0m")
}

func colour(red, green, blue float64) int {
	return 16 + baseColor(red, 36) + baseColor(green, 6) + baseColor(blue, 1)
}

func baseColor(value float64, factor int) int {
	return int(6*value/256) * factor
}
